src/pubsub.ts

```diff
@@ -57,56 +59,65 @@ export class AMQPPubSub implements PubSubEngine {
     if (refs && refs.length > 0) {
       const newRefs = [...refs, id];
       this.subsRefsMap[routingKey] = newRefs;
-      return Promise.resolve(id);
-    } else {
-      return this.subscriber.subscribe(this.exchange, routingKey, this.onMessage.bind(this))
-      .then(disposer => {
-        this.subsRefsMap[routingKey] = [
-          ...(this.subsRefsMap[routingKey] || []),
-          id,
-        ];
-        if (this.unsubscribeMap[routingKey]) {
-          return disposer();
-        }
-        this.unsubscribeMap[routingKey] = disposer;
-        return Promise.resolve(id);
-      });
+      return id;
     }
+
+    this.subsRefsMap[routingKey] = [
+      ...(this.subsRefsMap[routingKey] || []),
+      id,
+    ];
+
+    const existingDispose = this.unsubscribeMap[routingKey];
+    // Get rid of exisiting subscription while we get a new one.
+    const [newDispose] = await Promise.all([
+      this.subscriber.subscribe(this.exchange, routingKey, this.onMessage),
+      existingDispose ? existingDispose() : Promise.resolve(),
```

Comment:
Would you need to call existingDispose?
I mean if the case that there is nothing in the subRefsMap but existingDispose is still available there must be a dispose in progress.
Wouldn't this cause issues if the dispose code is run multiple times?

Reply:
I actually considered removing this code since I didn't see a case where it would be hit, but kept it to keep https://github.com/Surnet/graphql-amqp-subscriptions/pull/13/files#diff-97651bbb24c7d7b4cd79852e80437882L68-L70 in since it was added for some purpose, even though that code returned an `undefined` as the sub id.

As for dispose being run multiple times, the result of that dispose will be the same promise so the underlying channel consumer `cancel` call will only happen once, and all points that are watching that promise will receive the same resolved value. The only point I can see churn/leaks on is if unsubs/subs are happening in quick succession from an empty starting state for a routing key. In that case we may churn several consumer tags but should always end up in a state where we have one consumer tag in the end. The key to this is that Node is single threaded, and that we always modify `subscriptionMap` and `subMapRefs` BEFORE calling any `async`/promise logic.

I can remove this code and we can rely on the other call points to clean this up or leave it as is. I think existing concurrent test cases prove this out as well.

Reply:
Yes this code should not run at all.
I think I coded this in as a "just in case" thing.

Since you definitely looked at every line of code I have written and have the same opinion I think we could remove it.
I'll leave it up to you.

Reply:
I think I'm ok leaving it, as you said "Just in case".